pkg/controllers/grafana-operator/controller.go

```diff
@@ -151,108 +159,180 @@ func RegisterWithManager(mgr controllerruntime.Manager) error {
-		r.logger.Info("Updating Grafana Operator OperatorGroup")
-		operatorGroup.Spec = NewOperatorGroup().Spec
-		return r.k8sClient.Update(ctx, &operatorGroup)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcileError(err)
 	}
+
+	// create
 	if errors.IsNotFound(err) {
-		r.logger.Info("Creating Grafana Operator OperatorGroup")
-		return r.k8sClient.Create(ctx, NewOperatorGroup())
+		r.logger.Info("Creating Grafana OperatorGroup")
+		err := r.k8sClient.Create(ctx, NewOperatorGroup())
+		return creationResult(5*time.Second, err)
 	}
 
-	r.logger.Error(err, "error reconciling operator group")
-	return err
+	// update
+	r.logger.Info("Updating Grafana OperatorGroup")
+	desired := NewOperatorGroup().Spec
+	operatorGroup.Spec = desired
+	err = r.k8sClient.Update(ctx, &operatorGroup)
+	return updationResult(err)
 }
 
-func (r *reconciler) reconcileSubscription(ctx context.Context) error {
+func (r *reconciler) reconcileSubscription(ctx context.Context) reconcileResult {
 	key := types.NamespacedName{
 		Name:      subscriptionName,
 		Namespace: Namespace,
 	}
 	var subscription v1alpha1.Subscription
 	err := r.k8sClient.Get(ctx, key, &subscription)
-	if err == nil {
-		r.logger.Info("Updating Grafana Operator Subscription")
-		subscription.Spec = NewSubscription().Spec
-		return r.k8sClient.Update(ctx, &subscription)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcileError(err)
 	}
+
+	// create
 	if errors.IsNotFound(err) {
 		r.logger.Info("Creating Grafana Operator Subscription")
-		return r.k8sClient.Create(ctx, NewSubscription())
+		err := r.k8sClient.Create(ctx, NewSubscription())
+		return creationResult(15*time.Second, err)
 	}
 
-	return err
+	r.logger.Info("Updating Grafana Operator Subscription")
+	subscription.Spec = NewSubscription().Spec
+	err = r.k8sClient.Update(ctx, &subscription)
+	return updationResult(err)
 }
 
-func (r *reconciler) reconcileGrafana(ctx context.Context) error {
+func (r *reconciler) reconcileGrafana(ctx context.Context) reconcileResult {
+	key := types.NamespacedName{
+		Name:      grafanaName,
+		Namespace: Namespace,
+	}
+
 	var grafana integreatlyv1alpha1.Grafana
-	key := types.NamespacedName{Name: grafanaName, Namespace: Namespace}
 	err := r.k8sClient.Get(ctx, key, &grafana)
-
-	if err == nil {
-		r.logger.Info("Updating Grafana instance")
-		grafana.Spec = newGrafana().Spec
-		return r.k8sClient.Update(ctx, &grafana)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcileError(err)
 	}
 
+	// create
 	if errors.IsNotFound(err) {
 		r.logger.Info("Creating Grafana instance")
-		return r.k8sClient.Create(ctx, newGrafana())
+		err := r.k8sClient.Create(ctx, newGrafana())
+		return creationResult(15*time.Second, err)
 	}
 
-	return err
+	// update
+	r.logger.Info("Updating Grafana instance")
+	grafana.Spec = newGrafana().Spec
+	err = r.k8sClient.Update(ctx, &grafana)
+	return updationResult(err)
+}
+
+func (r *reconciler) reconcileGrafanaCRD(ctx context.Context) reconcileResult {
```

Comment:
Instead of adding additional RBAC on CRDs and adding this logic, could we simply requeue the request if we get an error that a resource does not exist?

Reply:
certainly worth a try. I will see if we can detect the `CRD not present` from a creation error.

Reply:
Without waiting for `CRD` to be `Established` we can get the following errors if `grafana` crd is deleted 
```
"error": "create not allowed while custom resource definition is terminating"}
```
The other errors like the `No matching kind` and `Not found` can be detected

Reply:
Did you mean can or _cannot_ be detected? If no matching kind is a detectable error, maybe that could be used to requeue the request?

Reply:
` errors.IsNotFound(err) || errors.IsMethodNotSupported(err) ` seems sufficient to wait for CRD creation.